flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/source/hybrid/HybridSourceReader.java

```diff
@@ -0,0 +1,213 @@
+ *
+ * <p>This reader is setup with a sequence of underlying source readers. At a given point in time,
+ * one of these readers is active. Underlying readers are opened and closed on demand as determined
+ * by the enumerator, which selects the active reader via {@link SwitchSourceEvent}.
+ *
+ * <p>When the underlying reader has consumed all input, {@link HybridSourceReader} sends {@link
+ * SourceReaderFinishedEvent} to the coordinator and waits for the {@link SwitchSourceEvent}.
+ */
+public class HybridSourceReader<T> implements SourceReader<T, HybridSourceSplit<?>> {
+    private static final Logger LOG = LoggerFactory.getLogger(HybridSourceReader.class);
+    private SourceReaderContext readerContext;
+    private List<SourceReader<T, ? extends SourceSplit>> realReaders;
+    private int currentSourceIndex = -1;
+    private long lastCheckpointId = -1;
+    private SourceReader<T, ? extends SourceSplit> currentReader;
+    // track last availability to resume reader after source switch
+    private CompletableFuture<Void> availabilityFuture;
+
+    public HybridSourceReader(
+            SourceReaderContext readerContext,
+            List<SourceReader<T, ? extends SourceSplit>> readers) {
+        this.readerContext = readerContext;
+        this.realReaders = readers;
+    }
+
+    @Override
+    public void start() {
+        setCurrentReader(0);
+    }
+
+    @Override
+    public InputStatus pollNext(ReaderOutput output) throws Exception {
+        InputStatus status = currentReader.pollNext(output);
+        if (status == InputStatus.END_OF_INPUT) {
+            // trap END_OF_INPUT if this wasn't the final reader
+            LOG.debug(
+                    "End of input subtask={} sourceIndex={} {}",
+                    readerContext.getIndexOfSubtask(),
+                    currentSourceIndex,
+                    currentReader);
+            if (currentSourceIndex + 1 < realReaders.size()) {
+                // Signal the coordinator that the current reader has consumed all input and the
+                // next source can potentially be activated (after all readers are ready).
+                readerContext.sendSourceEventToCoordinator(
+                        new SourceReaderFinishedEvent(currentSourceIndex, lastCheckpointId));
+                // More data will be available from the next reader.
+                // InputStatus.NOTHING_AVAILABLE requires us to complete the availability
+                // future after source switch to resume poll.
+                return InputStatus.NOTHING_AVAILABLE;
+            }
+        }
+        return status;
+    }
+
+    @Override
+    public List<HybridSourceSplit<?>> snapshotState(long checkpointId) {
+        this.lastCheckpointId = checkpointId;
+        List<? extends SourceSplit> state = currentReader.snapshotState(checkpointId);
+        return wrappedSplits(currentSourceIndex, state);
+    }
+
+    public static List<HybridSourceSplit<?>> wrappedSplits(
```

Comment:
Does the `wrappedSplits` update to `wrapSplits`?

Reply:
+1. action as method name